### ISSUE TITLE
[Bug] Xamarin.Forms.Entry does not enter decimal when binding a float/double and decimal to it

### ISSUE THREAD

**Issue by sunilpatil919** [xamarin/Xamarin.Forms#7996]
### Description
[Bug] Xamarin.Forms.Entry does not enter decimal when binding a float/double and decimal to it
### Steps to Reproduce

1. Take Entry. 
2.  Binding decimal variable to it. 
3.  if enter with one number and the enter decimal.
4. it not taking decimal after any number but first decimal then number it taking. 

### Basic Information
not working on Android and IOS

- Version with issue: XF-4.2.0.778463
- Last known good version:XF 3.6
- IDE:VS2017
- Platform Target Frameworks: 
  - iOS:  
  - Android:  
  - UWP:   
- Android Support Library Version: 
- Nuget Packages:
- Affected Devices:

**Comment by adrianknight89** [comment]
@sunilpatil919 Would be great if you could attach a minimal reproduction project to expedite verification.

**Comment by samhouts** [comment]
@sunilpatil919  Can you please attach a small project that demonstrates this issue? Thanks!

**Comment by samhouts** [comment]
@sunilpatil919  Since we haven't heard from you in more than 30 days, we hope this issue is no longer affecting you. If it is, please reopen this issue and provide the requested information so that we can look into it further. Thank you!

**Comment by E75** [comment]
@samhouts 
Indeed, this doesn't work on X.Forms 4.2+ versions. Only X.Forms 4.1.0.778454 and under works great

**Comment by samhouts** [comment]
@E75  Can you please attach a small project that demonstrates this issue? Thanks!

**Comment by E75** [comment]
@samhouts 
I see that this occurs, when I use ObservableCollection... 
X.Forms 4.1.0.778454 and lower works great. With the latest update, this doesn't work anymore
This doesn't also work in UWP example...

Sample project. => Use UWP example project
[EntryBug.zip](https://github.com/xamarin/Xamarin.Forms/files/3876696/EntryBug.zip)

**Comment by WayaFlyfeather** [comment]
@E75, in the example you provided, this behavior is caused by the code not checking if the new value is different from previous value in the EntryBugModel.Value attribute setter.

What happens is, the value is 5 (from initialization). User enters '.' after the 5. EntryBugModel.Value is called, and because there is no check if 5 is different from 5, PropertyChanged is raised, which in turn resets the field to 5 (without the '.').

If you change the setter to:
```
            set
            {
                if (value != _value)
                {
                    _value = value;

                    OnPropertyChanged("Value");
                }
            }
```
it works fine.

**Comment by WayaFlyfeather** [comment]
(If the code worked in previous versions, it may indicate that a check in the XF binding code, for if values are equal, has been removed. Not sure if this would be a bug or not.)

**Comment by jonkas2211** [comment]
I have found out that the error is cause by the `BindingExpression.TryConvert` method.

The method contains the following call `value = Convert.ChangeType(value, convertTo, CultureInfo.InvariantCulture);`
I tested this on german culture so value = "1,5", convertTo = decimal Type. When I change the CultureInfo.InvariantCulture to CultureInfo.CurrentUICulture the conversion is made correctly. 

**BUT** i can not enter 1,5 easily. When i write the ',' nothing will happen, because from 1, will get to a 1.

**Comment by AdrianoRochedo** [comment]
This must be tested for many international languages as possible .. ever !!!
Look as the value 105.69 gets with the folow string formats (using pt-br)

![image](https://user-images.githubusercontent.com/46300640/108549504-f9cfba00-72cb-11eb-8bea-4c700a8dd79c.png)

![image](https://user-images.githubusercontent.com/46300640/108549599-1ec42d00-72cc-11eb-96b6-57f92c315d85.png)

**Comment by Vroomer** [comment]
Don't worry about this issue. It's only a half of the World using commas as a decimal separator. I'm sure it can wait another year. https://en.wikipedia.org/wiki/Decimal_separator#/media/File:DecimalSeparator.svg

**Comment by emilsteen** [comment]
Before any automatic triage closes this because there is no activity I'll add an update:
The issue is still in Visual Studio 16.10-preview2.0.

**Comment by Lebapp** [comment]
things like this make it hard to stay with xamarin or even recommend it

**Comment by AndrewHoUs** [comment]
It only require to charge 2 lines of code, why this is not fixed yet?
[first line](https://github.com/xamarin/Xamarin.Forms/blob/78f4012603efcb1803d5b1cc3baecdde2a497096/Xamarin.Forms.Core/BindingExpression.cs#L463)
[second line](https://github.com/xamarin/Xamarin.Forms/blob/78f4012603efcb1803d5b1cc3baecdde2a497096/Xamarin.Forms.Core/BindingExpression.cs#L478)

**Comment by ilsur-dev111** [comment]
I also ran into this problem, there is even a solution to why the problem is still not solved?

**Comment by AdrianoRochedo** [comment]
Its a Shame !

Em qua, 1 de dez de 2021 07:34, ilsur-dev111 ***@***.***>
escreveu:

**Comment by francescoformenti** [comment]
> I also ran into this problem, there is even a solution to why the problem is still not solved?

I found a quite simple workaround using a value converter in the binding:
```
public class DecimalNumberWorkaroundConverter : IValueConverter
{

	public object Convert(object value, Type targetType, object parameter, CultureInfo culture)
	{
		return string.Format(culture ?? CultureInfo.CurrentCulture, "{0}", value);
	}

	public object ConvertBack(object value, Type targetType, object parameter, CultureInfo culture)
	{
		var selectedCulture = culture ?? CultureInfo.CurrentCulture;
		if (ToNullableDecimal(value?.ToString(), selectedCulture) is decimal decValue)
		{
			return System.Convert.ChangeType(decValue, targetType, selectedCulture);
		}
		else
		{
			return value;
		}
	}

	private static decimal? ToNullableDecimal(string value, IFormatProvider provider)
	{
		if (string.IsNullOrEmpty(value))
		{
			try
			{
				if (decimal.TryParse(value, NumberStyles.Any, provider, out decimal result))
				{
					return result;
				}
			}
			catch { }
		}
		return null;
	}

}
```

**Comment by jfversluis** [comment]
Hey everyone! Yes, this should have been fixed a long time ago and yes it's a shame that we left the PR that fixes this linger around for so long, so sorry about that. There is nothing I can change about that any more, but what I _can_ change is what happens with it now; so let's fix this!

I've reactivated the PR (many thanks to @jonkas2211 for the initial effort!) and it looks like the code is good, just a unit test that needed updating. I did that now and the build is running. If someone still watching this and waiting, would you please be so kind to grab the NuGet from #11815 [as described here](https://github.com/xamarin/Xamarin.Forms/wiki/Preview-Packages) and let us know if this indeed fixes this issue? That will greatly speed up the review process.

Besides verifying if this particular issue is fixed also be sure to check other scenarios in the same area to make sure that this fix doesn't accidentally has side-effects 🙂 

Thanks!
